tests/test_validation.py

```diff
@@ -0,0 +1,184 @@
+    ])
+
+    resp = app.post('/report', json={
+        'type': 'report',
+        'report_type': 'simple',
+        'status': 'valid',
+        'valid_from_date': '2019-04',  # invalid conversion to date
+    })
+
+    assert resp.status_code == 400
+    assert resp.json() == {
+        "error": "Invalid isoformat string: '2019-04'",
+    }
+
+
+def test_non_string_report_date(context, app):
+    app.authorize([
+        'spinta_report_insert',
+    ])
+
+    resp = app.post('/report', json={
+        'type': 'report',
+        'report_type': 'simple',
+        'status': 'valid',
+        'valid_from_date': 42,  # invalid conversion to date
+    })
+
+    assert resp.status_code == 400
+    assert resp.json() == {
+        "error": "fromisoformat: argument must be str",
+    }
+
+
+def test_invalid_report_datetime(context, app):
+    app.authorize([
+        'spinta_report_insert',
+    ])
+
+    resp = app.post('/report', json={
+        'type': 'report',
+        'report_type': 'simple',
+        'status': 'valid',
+        'update_time': '2019-04',  # invalid conversion to datetime
+    })
+
+    assert resp.status_code == 400
+    assert resp.json() == {
+        "error": "Invalid isoformat string: '2019-04'",
+    }
+
+
+def test_non_string_report_datetime(context, app):
+    app.authorize([
+        'spinta_report_insert',
+    ])
+
+    resp = app.post('/report', json={
+        'type': 'report',
+        'report_type': 'simple',
+        'status': 'valid',
+        'update_time': 42,  # invalid conversion to datetime
+    })
+
+    assert resp.status_code == 400
+    assert resp.json() == {
+        "error": "fromisoformat: argument must be str",
+    }
+
+
+def test_invalid_report_array(context, app):
+    app.authorize([
+        'spinta_report_insert',
+    ])
+
+    resp = app.post('/report', json={
+        'type': 'report',
+        'report_type': 'simple',
+        'status': 'valid',
+        'notes': {'foo': 'bar'},  # invalid conversion to array
+    })
+
+    assert resp.status_code == 400
+    assert resp.json() == {
+        "error": "Invalid array type: <class 'dict'>",
+    }
+
+
+def test_invalid_report_array_object(context, app):
+    app.authorize([
+        'spinta_report_insert',
+    ])
+
+    resp = app.post('/report', json={
+        'type': 'report',
+        'report_type': 'simple',
+        'status': 'valid',
+        'notes': ['hello', 'world'],  # invalid array item type
+    })
+
+    assert resp.status_code == 400
+    assert resp.json() == {
+        "error": "Invalid object type: <class 'str'>",
+    }
+
+
+def test_invalid_nested_object_property(context, app):
+    app.authorize([
+        'spinta_nested_insert',
+    ])
+
+    resp = app.post('/nested', json={
+        'some': [{
+            'nested': 'object' # invalid object property type
+        }]
+    })
+
+    assert resp.status_code == 400
+    assert resp.json() == {
+        "error": "Invalid object type: <class 'str'>",
+    }
```

Comment:
I would like to see a test, where you pass unknown properties. Not sure if checks for required properties, that would be good to test too.

Reply:
Also, add tests when nested dict does not have a known property.

Reply:
> Not sure if checks for required properties, that would be good to test too.

I have not done any logic to check for property's properties such as required.